File: clients/go/model/vehicle.go
package model

import "fmt"
import "io"
import . "komus24/stream"

// TODO - Document
type Vehicle struct {
    // TODO - Document
    Position Vec2Float64
    // TODO - Document
    Velocity Vec2Float64
    // TODO - Document
    Speed float64
    // TODO - Document
    RotationSpeed float64
    // TODO - Document
    Rotation float64
    // TODO - Document
    TypeIndex int32
    // TODO maybe multiple quests at the same time?
    Quest *Quest
    // TODO - Document
    Fuel float64
}

func NewVehicle(position Vec2Float64, velocity Vec2Float64, speed float64, rotationSpeed float64, rotation float64, typeIndex int32, quest *Quest, fuel float64) Vehicle {
    return Vehicle {
        Position: position,
        Velocity: velocity,
        Speed: speed,
        RotationSpeed: rotationSpeed,
        Rotation: rotation,
        TypeIndex: typeIndex,
        Quest: quest,
        Fuel: fuel,
    }
}

// Read Vehicle from reader
func ReadVehicle(reader io.Reader) Vehicle {
    var position Vec2Float64
    position = ReadVec2Float64(reader)
    var velocity Vec2Float64
    velocity = ReadVec2Float64(reader)
    var speed float64
    speed = ReadFloat64(reader)
    var rotationSpeed float64
    rotationSpeed = ReadFloat64(reader)
    var rotation float64
    rotation = ReadFloat64(reader)
    var typeIndex int32
    typeIndex = ReadInt32(reader)
    var quest *Quest
    if ReadBool(reader) {
        var questValue Quest
        questValue = ReadQuest(reader)
        quest = &questValue
    } else {
        quest = nil
    }
    var fuel float64
    fuel = ReadFloat64(reader)
    return Vehicle {
        Position: position,
        Velocity: velocity,
        Speed: speed,
        RotationSpeed: rotationSpeed,
        Rotation: rotation,
        TypeIndex: typeIndex,
        Quest: quest,
        Fuel: fuel,
    }
}

// Write Vehicle to writer
func (vehicle Vehicle) Write(writer io.Writer) {
    position := vehicle.Position
    position.Write(writer)
    velocity := vehicle.Velocity
    velocity.Write(writer)
    speed := vehicle.Speed
    WriteFloat64(writer, speed)
    rotationSpeed := vehicle.RotationSpeed
    WriteFloat64(writer, rotationSpeed)
    rotation := vehicle.Rotation
    WriteFloat64(writer, rotation)
    typeIndex := vehicle.TypeIndex
    WriteInt32(writer, typeIndex)
    quest := vehicle.Quest
    if quest == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        questValue := *quest
        questValue.Write(writer)
    }
    fuel := vehicle.Fuel
    WriteFloat64(writer, fuel)
}

// Get string representation of Vehicle
func (vehicle Vehicle) String() string {
    stringResult := "{ "
    stringResult += "Position: "
    position := vehicle.Position
    stringResult += position.String()
    stringResult += ", "
    stringResult += "Velocity: "
    velocity := vehicle.Velocity
    stringResult += velocity.String()
    stringResult += ", "
    stringResult += "Speed: "
    speed := vehicle.Speed
    stringResult += fmt.Sprint(speed)
    stringResult += ", "
    stringResult += "RotationSpeed: "
    rotationSpeed := vehicle.RotationSpeed
    stringResult += fmt.Sprint(rotationSpeed)
    stringResult += ", "
    stringResult += "Rotation: "
    rotation := vehicle.Rotation
    stringResult += fmt.Sprint(rotation)
    stringResult += ", "
    stringResult += "TypeIndex: "
    typeIndex := vehicle.TypeIndex
    stringResult += fmt.Sprint(typeIndex)
    stringResult += ", "
    stringResult += "Quest: "
    quest := vehicle.Quest
    if quest == nil {
        stringResult += "nil"
    } else {
        questValue := *quest
        stringResult += questValue.String()
    }
    stringResult += ", "
    stringResult += "Fuel: "
    fuel := vehicle.Fuel
    stringResult += fmt.Sprint(fuel)
    stringResult += " }"
    return stringResult
}

File: clients/rust/src/model/vehicle.rs
use super::*;

/// A vehicle
#[derive(Clone, Debug)]
pub struct Vehicle {
    /// Current position (center)
    pub position: model::Vec2F64,
    /// Velocity vector
    pub velocity: model::Vec2F64,
    /// Speed of wheels
    pub speed: f64,
    /// Rotation speed (radians/second)
    pub rotation_speed: f64,
    /// Current rotation
    pub rotation: f64,
    /// Vehicle type index
    pub type_index: i32,
    /// Current quest, if any
    pub quest: Option<model::Quest>,
    /// Remaining fuel
    pub fuel: f64,
}

impl trans::Trans for Vehicle {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.position.write_to(writer)?;
        self.velocity.write_to(writer)?;
        self.speed.write_to(writer)?;
        self.rotation_speed.write_to(writer)?;
        self.rotation.write_to(writer)?;
        self.type_index.write_to(writer)?;
        self.quest.write_to(writer)?;
        self.fuel.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let position: model::Vec2F64 = trans::Trans::read_from(reader)?;
        let velocity: model::Vec2F64 = trans::Trans::read_from(reader)?;
        let speed: f64 = trans::Trans::read_from(reader)?;
        let rotation_speed: f64 = trans::Trans::read_from(reader)?;
        let rotation: f64 = trans::Trans::read_from(reader)?;
        let type_index: i32 = trans::Trans::read_from(reader)?;
        let quest: Option<model::Quest> = trans::Trans::read_from(reader)?;
        let fuel: f64 = trans::Trans::read_from(reader)?;
        Ok(Self {
            position,
            velocity,
            speed,
            rotation_speed,
            rotation,
            type_index,
            quest,
            fuel,
        })
    }
}